### PR TITLE
Parse multiple volumes via semicolon delimeter.

### PR DESCRIPTION
**Pull request by adambkaplan**
Fixes issue #872

### CODE REVIEW

File: hack/test-stirunimage.sh [hunk 1]
```diff
@@ -140,6 +140,10 @@ test_debug "s2i build with remote git repo"
 s2i build https://github.com/openshift/cakephp-ex docker.io/centos/php-70-centos7 test --loglevel=5 &> "${WORK_DIR}/s2i-git-proto.log"
 check_result $? "${WORK_DIR}/s2i-git-proto.log"
 
+test_debug "s2i build with runtime image"
+s2i build --ref=10.x --context-dir=helloworld https://github.com/wildfly/quickstart docker.io/openshift/wildfly-101-centos7 test-jee-app-thin --runtime-image=docker.io/openshift/wildfly-101-centos7 &> "${WORK_DIR}/s2i-runtime-image.log"
```
Comment (adambkaplan) [inline hack/test-stirunimage.sh]: Note - the wildfly s2i images were recently updated to support runtime artifacts/builds.

File: pkg/api/types.go [hunk 2]
```diff
@@ -527,8 +527,22 @@ func IsInvalidFilename(name string) bool {
...
+		newVols[i] = spec
+	}
+	*l = append(*l, newVols...)
+	return nil
+}
+
+func (l *VolumeList) parseSpec(value string) (VolumeSpec, error) {
 	if len(value) == 0 {
-		return errors.New("invalid format, must be source:destination")
+		return VolumeSpec{}, errors.New("invalid format, must be source:destination")
```
Comment (gabemontero) [inline pkg/api/types.go]: should return nil vs. empty

File: pkg/api/types.go [hunk 3]
```diff
@@ -541,10 +555,9 @@ func (l *VolumeList) Set(value string) error {
 	mount[1] = strings.Trim(mount[1], `"'`)
 	s := VolumeSpec{Source: filepath.Clean(mount[0]), Destination: filepath.ToSlash(filepath.Clean(mount[1]))}
 	if IsInvalidFilename(s.Source) || IsInvalidFilename(s.Destination) {
-		return fmt.Errorf("invalid characters in filename: %q", value)
+		return VolumeSpec{}, fmt.Errorf("invalid characters in filename: %q", value)
```
Comment (gabemontero) [inline pkg/api/types.go]: should return nil vs. empty

**Comment by gabemontero** [comment]
[test]

**Comment by gabemontero** [comment]
I think s2i still leverages rosie ... let's see

**Comment by gabemontero** [comment]
/test

maybe not rosie

**Comment by gabemontero** [comment]
No I was right ... still rosie ... there she goes

File: pkg/api/types.go (hunk 2 above)
Comment (adambkaplan) [reply pkg/api/types.go]: Can't do that for value return :/

Comment (gabemontero) [reply pkg/api/types.go]: sorry, the implication was to change the return type to a pointer

Comment (gabemontero) [reply pkg/api/types.go]: unless there are errors that can be selectively ignored by the caller, why return an empty or partial VolumeSpec ... less space on the stack as well

**Comment by adambkaplan** [comment]
[test]

**Comment by bparees** [comment]
this lgtm but one concern:

what's the behavior in origin today if someone specifies a path uses a ;?  Are we erroring out, or are we tolerating the filename/path as being "some;path"?

Because if we're tolerating it as a valid filename today, this PR will change that behavior in origin and that would be a breaking change.

**Comment by adambkaplan** [comment]
Based on the prior state, s2i rejects `some;path` as an illegal file name - `;` is an invalid file name character. See [here](https://github.com/openshift/source-to-image/blob/master/pkg/api/types.go#L26)

**Comment by bparees** [comment]
ok, seems like this should be safe then, thanks.

[merge]

**Pull request merged**